architecture/2024-05-radius-on-dapr.md

```diff
@@ -0,0 +1,324 @@
+
+The storage API in Radius is already oriented around key-value stores, so it's mostly a 1:1 replacement. 
+
+The challenging portion is that Dapr lacks a widely-supported query API. The state store building block includes an optional query feature, but it's deprecated and not widely supported. We shouldn't use it.
+
+The alternative is that we build our own "index" feature when inserting or deleting records. This would be used when querying resources at a given scope. 
+
+If we use the resource ID as the state store key, an example scheme might look something like this:
+
+```txt
+/planes/radius/local/resourcegroups/my-app/providers/applications.core/containers/frontend -> (data for frontend)
+/planes/radius/local/resourcegroups/my-app/providers/applications.core/containers/backend -> (data for backend)
+/planes/radius/local/resourcegroups/my-app/:index/applications.core -> (list of container resource names)
+/planes/radius/local/resourcegroups -> (list of container resource names)
+```
+
+This allows us to implement query support with two roundrips to the database. First, we fetch the index, then we fetch the resources using the names in the index.
+
+#### Pub/Sub + Async Worker
+
+Our task is to reimplement the [async worker](https://github.com/radius-project/radius/blob/main/pkg/armrpc/asyncoperation/worker/worker.go#L130) using Dapr pub/sub instead of our message queue abstraction. Since we'll be keeping the existing controller API this change should have no impact on the rest of the codebase.
+
+There's an important [semantic difference](https://www.baeldung.com/pub-sub-vs-message-queues) between a message queue and the kind of pub/sub implemented by Dapr. Our current async worker implementation relies heavily on the *extend* and *retry* semantics that are common in message queues. Basically when we read a message, we don't *commit* the read until it's been processed end-to-end, which might be an hour later. This isn't a pattern that pub/sub supports.
+
+To address this, we'll use pub/sub messages to trigger an actor, and use the actor to execute the existing logic that applies retries and backoff. This ensures that the message is not *committed* until it's been accepted by an actor. Actors provide their own functionality for deferred retries that we can leverage. 
+
+-- 
+
+I'm also working on an additional proposal to replace the async worker with Dapr workflows. If we think it's better, we could skip the intermediate step of upgrading the current architecture. The path outlined above is the most incremental approach. 
+
+#### Development Experience
+
+Since Dapr is a new required component, we will need to make some investments or our development experience will be poor. 
+
+I'm proposing that we require Dapr to by installed by users to develop Radius. We'll need the `daprd` binary to be present on the user's system.
```

Comment:
Should a user be given an option to not install Dapr and do it the way we have now? Do we need to be heavily dependent on Dapr?

Reply:
We either take a dependency or not. We can't have the benefits without being required, because we'll be building an abstraction layer for an abstraction layer.